### PR TITLE
core[patch]: Reducing heap area consumption regardless of the number of prompts

### PR DESCRIPTION
**Pull request by tsukasaoishi**
Currently, the memory heap area consumption increases proportionally with the number of prompt messages. This pull request addresses this issue, ensuring the heap area usage remains minimal regardless of the number of messages.

To illustrate this, I have made modifications to the example code ([`examples/src/chat/llm_chain.ts`](https://github.com/langchain-ai/langchainjs/blob/main/examples/src/chat/llm_chain.ts)) as shown below:

```
import { LLMChain } from "langchain/chains";
import { ChatOpenAI } from "langchain/chat_models/openai";
import { ChatPromptTemplate } from "langchain/prompts";

const memoryCheck = () => {
  const used = process.memoryUsage()
  const messages = []
  for (let key in used) {
    messages.push(`${key}: ${Math.round(used[key] / 1024 / 1024 * 100) / 100} MB`)
  }
  console.log(new Date(), messages.join(', '))
}
  
export const run = async () => {
  const chat = new ChatOpenAI({ temperature: 0 });

  const chatPrompt = ChatPromptTemplate.fromMessages([
    [
      "system",
      "You are a helpful assistant that translates {input_language} to {output_language}.",
    ],
    ["human", "{text}"],
    ["human", "{text}"],
    ["human", "{text}"],
    ["human", "{text}"],
    ["human", "{text}"],
    ["human", "{text}"],
    ["human", "{text}"],
    ["human", "{text}"],
    ["human", "{text}"],
    ["human", "{text}"],
    ["human", "{text}"],
    ["human", "{text}"],
    ["human", "{text}"],
    ["human", "{text}"],
    ["human", "{text}"],
    ["human", "{text}"],
    ["human", "{text}"],
    ["human", "{text}"],
    ["human", "{text}"],
    ["human", "{text}"],
  ]);

  const chain = new LLMChain({
    prompt: chatPrompt,
    llm: chat,
  });

  console.log("start");
  memoryCheck();
  const response = await chain.call({
    input_language: "English",
    output_language: "French",
    text: "I love programming.",
  });
  console.log("end");
  memoryCheck();

  console.log(response);
};
```

Upon execution, the heap area usage is as follows:

```
start
2023-12-04T06:10:43.821Z rss: 110.97 MB, heapTotal: 59.7 MB, heapUsed: 30.18 MB, external: 1.35 MB, arrayBuffers: 0.44 MB
end
2023-12-04T06:13:31.299Z rss: 1104.91 MB, heapTotal: 978.66 MB, heapUsed: 939.35 MB, external: 2.46 MB, arrayBuffers: 0.25 MB
```

Although these memories are released when the Garbage Collector (GC) runs, there is a temporary bloating of memory. If the allocated memory is insufficient, this can lead to a memory shortage in the heap area, resulting in an error. This error is particularly frequent in environments with limited memory, such as Kubernetes.

The root cause of this issue is the creation of a Tiktoken object for each prompt. A single Tiktoken object can consume close to 50MB of memory, leading to increased memory consumption as the number of prompts grows.

After applying this pull request, the heap area consumption for the same code is as follows:

```
start
2023-12-04T06:23:46.842Z rss: 108.02 MB, heapTotal: 58.7 MB, heapUsed: 32.56 MB, external: 1.37 MB, arrayBuffers: 0.45 MB
end
2023-12-04T06:23:48.079Z rss: 216.45 MB, heapTotal: 104.34 MB, heapUsed: 74.17 MB, external: 2.34 MB, arrayBuffers: 0.07 MB
```

### twitter
[tsukasa_oishi](https://twitter.com/tsukasa_oishi)

### CODE REVIEW

**Comment by jacoblee93** [comment]
Wow neat fix - CC @dqbd

File: langchain-core/src/utils/tiktoken.ts [hunk 1]
```diff
@@ -1,44 +1,34 @@
...
-const cache: Record<string, Promise<TiktokenBPE>> = {};
+const cache: Record<string, Promise<Tiktoken>> = {};
 
 const caller = /* #__PURE__ */ new AsyncCaller({});
 
 export async function getEncoding(
   encoding: TiktokenEncoding,
-  options?: {
-    signal?: AbortSignal;
-    extendedSpecialTokens?: Record<string, number>;
```
Comment (jacoblee93) [inline langchain-core/src/utils/tiktoken.ts:18]: Were we never using these properties?

Comment (dqbd) [reply langchain-core/src/utils/tiktoken.ts:18]: Seems like we're only using for estimate token counts in `BaseLanguageModel::getNumTokens`, which does not handle AbortSignal as well.

**Review by dqbd — APPROVED** [review]

**Comment by jacoblee93** [comment]
Thank you very much!

**Pull request merged**